packages/paima-sdk/paima-events/src/event-listener.ts

```diff
@@ -1,60 +1,118 @@
 /* eslint-disable no-console */
+import { EventPathAndDef, fillPath, ResolvedPath, Undefined } from './builtin-events.js';
 import { PaimaEventConnect } from './event-connect.js';
-import type { PaimaEvent } from './events.js';
+import { PaimaEventBrokerNames } from './event-utils.js';
+import type { Static } from '@sinclair/typebox';
 
-export type MQTTCallback = (
-  target: string,
-  path: string | ((address: string) => string),
-  rawObject: object
-) => void;
+export type CallbackArgs<Event extends EventPathAndDef> = {
+  // schema of the content emitted
+  // ex: block/5 -> block content
+  val: Static<Event['type']>;
+  // resolution of the variables specified in the topic name
+  // ex: block/5 -> { blockId: 5 }
+  resolvedPath: ResolvedPath<Event['path']>;
+};
 
+export type CallbackAndMetadata<Event extends EventPathAndDef> = {
+  callback: (args: CallbackArgs<Event>) => void;
+  event: Event;
+};
 
-/* 
+/*
  * This class subscribes to specific topics, and stores the callbacks for event processing.
  */
 export class PaimaEventListener {
```

Comment:
Now that this class works as Listener and Publisher, maybe rename just to PaimaEvents